Zeromq transport python3

thatch45 opened the pull request:
### What does this PR do?

Brings python3 to the zeromq transport, so now we have tcp and zmq
### What issues does this PR fix or reference?
### Previous Behavior

Remove this section if not relevant
### New Behavior

Remove this section if not relevant
### Tests written?

Yes/No

thatch45 committed:
Zeromq transport python3

thatch45 commented:
Thanks @skizunov! You did the heavy lifting!

@skizunov, how far have you tested python3 on the tcp transport and windows? I can do a test.ping, but states fail, so I am just curious how far you got before I start ripping deeper

thatch45 committed:
lets toss this fix in here, so that we get exception messages

skizunov commented:
We don't (yet) use states. We use the job system to execute modules on the salt-minion. So I tested such use cases with our software stack.

On my system, the salt-master and salt-minion start up cleanly. I am able to correctly execute jobs (via custom modules) on the salt-minion via the `salt` command run from the salt-master. I also tested connect/reconnect cases for taking the master or minion down and back up again (presence_events). We have our own `master_job_cache` so I haven't tried using the default one. We make use of the reactor system for translating events to runner calls and so that seems to be working as well.

For our engine to run, I needed to change our conf file from this:

```
engines:
  - nisms
```

to this:

```
engines: {
  nisms: {}
}
```

I will probably look into fixing that sooner or later since I think the first style is cleaner. You may see some minor fixes come in from me for various miscellaneous things in the future. Once we have internal builds of our stack using Python 3 (hopefully in a few weeks) and are able to run our own test suites, we will likely identify more problems for which I will post fixes. I have other tasks outside of Salt so it is not certain when I can find time to do so.

As you may have guessed, one of my priorities is to implement the load balancer for TCP transport on Windows. The Windows salt-master is a bit crippled without it. The PY3 support is actually the first step towards that goal.

thatch45 commented:
Thanks @skizunov, this helps a great deal. There are a bunch of places I am seeing that still need work, but they all are outside of what you are working on, like the minion proc dir and the salt fileserver, so I will keep pushing down the road I am on!

skizunov commented:
The intention behind my submission was to get Python 3 working for the exact configuration and use cases that my company uses. It was never intended to get the whole range of product features working with Python 3. Hopefully my submission is a good start for your work.

thatch45 commented:
It is a fantastic start! With this we are actually very close, I just wanted to make sure I knew where you were at before I started galumphing around in PY3 land

thatch45 committed:
got states working

This forces the serializer to deserialize into strings on py3
but also allows for the argument 'raw' to be passed in allowing
for raw bytes to be sent explicitly. I made it this way
because the majority of calls are operational and the fileclient
is the only thing that needs bytes from the master

thatch45 commented:
OK, I have pushed new code to this branch but it is not showing up in github, this fills me with fear as to what would be merged if I pushed the merge button now, so wait until the added get states working in py3 is in here

thatch45 commented:
Ok, my additional code is in here now, so if tests pass and review passes this should be good to go

s0undt3ch commented:
Hmm....

`str()` should probably call `__unicode__()` so it should be safe...

thatch45 committed:
all channels need to absorb the new raw arg

skizunov commented on salt/payload.py:
```diff
@@ -132,9 +133,11 @@ def loads(self, msg, encoding=None):
-                return msgpack.loads(msg, use_list=True)
+                ret = msgpack.loads(msg, use_list=True)
+            if six.PY3 and not raw:
+                ret = salt.transport.frame.decode_embedded_strs(ret)
```
The original design had the caller of `loads` invoke `decode_embedded_strs`. If you add `decode_embedded_strs(ret)` here, then you should remove it from the callers (eg in 'transport/tcp.py' the callers that use `salt.crypt.Cryptical.loads` but not the ones that use the `msgpack.Unpacker`). Double decoding is inefficient. Or you could continue the tradition of having the caller responsible for this type of decoding. Your call.

Also, you don't want to run `decode_embedded_strs` when encoding is not None. The reason is that if encoding is not None, then it has been encoded with `use_bin_type=True`, in which case msgpack has already saved the distinction between `bytes` and `str`. `use_bin_type=True` is used in the IPC and events case since we don't care about breaking the wire protocol in those cases and it is much more efficient to have msgpack properly encode things in the first place.

eg the conditional would become:

`if six.PY3 and encoding is None and not raw:`

thatch45 replied:
Thanks, I will get that in

thatch45 replied:
I debated a lot about this, and went through all of the sale code and concluded that it would be better this way, the only problem I ran into was the file client which needs the data in bytes, all other data was operational and so I felt it was ok to assume all operational data can be decoded. Also, the only really large data sets are in the file client, so avoiding them there is also wise. And yes, I don't want to double decode, I just need to clean up some more.

thatch45 replied:
Also, as for the tcp transport, it calls msgpack directly (which it should not do) so it would still need to decode, I will look at cleaning that up in another PR

skizunov commented on salt/crypt.py:
```diff
@@ -1239,7 +1239,7 @@ def loads(self, data):
-        if six.PY3:
+        load = self.serial.loads(data[len(self.PICKLE_PAD):], raw=raw)
+        if six.PY3 and not raw:
             load = salt.transport.frame.decode_embedded_strs(load)
```
You changed `self.serial.loads` to call `decode_embedded_strs`. So this logic would double-decode which is inefficient.

skizunov commented on salt/fileclient.py:
```diff
@@ -956,31 +956,31 @@ def get_file(self,
                         d_tries += 1
-                        hsum = salt.utils.get_hash(dest, data.get('hash_type', 'md5'))
-                        if hsum != data['hsum']:
+                        hsum = salt.utils.get_hash(dest, data.get(b'hash_type', 'md5'))
```
Since you are using 'raw' mode, if 'hash_type' is in data, then its value will be stored as bytes instead of str. I believe that `salt.utils.get_hash` currently expects that arg to be a str (since it uses `getattr` and the docs say the name must be str). So perhaps you will need something like:

```
hsum = salt.utils.get_hash(dest, salt.utils.to_str(data.get(b'hash_type', b'md5')))
```

thatch45 replied:
A very nice catch, but this passed my usage tests. I will take a closer look, thanks

thatch45 committed:
Soem redundancy cleanup

thatch45 committed:
fix up base class

thatch45 commented:
Hmm, still getting that error, I thought I had updated all the send refs, I am looking again

thatch45 committed:
try some more...

thatch45 commented:
Go Go Jenkins!

skizunov committed:
Fix for "TypeError: send() got an unexpected keyword argument 'raw'"

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

thatch45 committed:
Merge pull request #9 from skizunov/zmqpy3

Fix for "TypeError: send() got an unexpected keyword argument 'raw'"

thatch45 commented:
@skizunov, that did it! Thanks for tracking down the one I missed!
We should get this merged in soon, then I will try to run the test suite on py3 and see how bad it is
